### PR TITLE
Add stronger typing for payload header APIs

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
#### Problem
Payload header API expects both a protocol id and a message type to be provided.  These can end up out of sync (e.g. a message type for one protocol can get used with a different protocol).

 #### Summary of Changes
Define a set of APIs that allow us to just pass in a strongly typed (enum class) message type and derive the protocol from the type of the message type.

Draft for now because it depends on https://github.com/project-chip/connectedhomeip/pull/4428 to pass tests and includes that changeset for the moment.

### CODE REVIEW

**Comment by bzbarsky-apple** [comment]
I guess I need to either figure out how to make the NRFConnect SDK test build use C++14 or stop using `std::underlying_type_t` (and use `std::underlying_type<>::type` instead).

**Comment by bzbarsky-apple** [comment]
Note: the size increase seems to be from the change to C++14 ...

**Comment by bzbarsky-apple** [comment]
https://github.com/project-chip/connectedhomeip/pull/4436 fixes the nrfconnect bits.

File: src/transport/raw/MessageHeader.h [hunk 1]
```diff
@@ -340,24 +352,37 @@ class PayloadHeader
         return *this;
     }
 
-    /** Set the secure message type for this header. */
-    PayloadHeader & SetMessageType(uint8_t type)
+    /**
+     * Set the message type for this header.  This requires setting the protocol
+     * id as well, because the meaning of a message type is only relevant given
+     * a speciifc protocol.
```
Comment (mspang) [inline src/transport/raw/MessageHeader.h:358]: typo: specific

**Review by mspang — APPROVED** [review]

**Review by rwalker-apple — APPROVED** [review]

**Review by jelderton — APPROVED** [review]

File: src/transport/raw/MessageHeader.h (hunk 1 above)
Comment (bzbarsky-apple) [reply src/transport/raw/MessageHeader.h]: Good catch, fixed.

**Pull request merged**